File: intel-hex-to-bin.java/src/main/java/demo/Memory.java
package demo;

/**
 * Memory image.
 */
final public class Memory {

    private short[] source = new short[0];

    public void expand(int address, short pad) {
        short[] source2 = new short[address];

        System.arraycopy(source, 0, source2, 0, source.length);

        for (int i = source.length; i < address; i++) {
            source2[i] = pad;
        }

        source = source2;
    }

    public void set(int index, short value) {
        source[index] = value;
    }

    public byte[] toArray() {
        byte[] target = new byte[source.length];
        for (int i = 0; i < source.length; i++) {
            target[i] = (byte) source[i];
        }
        return target;
    }

    public int size() {
        return source.length;
    }

}

File: intel-hex-to-bin.kotlin/src/main/kotlin/demo/Memory.kt
package demo

/**
 * Memory image.
 */
class Memory {

    private var source = ShortArray(0)

    fun expand(address: Int, pad: Short) {
        val source2 = ShortArray(address)

        System.arraycopy(source, 0, source2, 0, source.size)

        for (i in source.size..address - 1) {
            source2[i] = pad
        }

        source = source2
    }

    fun set(index: Int, value: Short) {
        source[index] = value
    }

    fun toArray(): ByteArray {
        val target = ByteArray(source.size)
        for (i in source.indices) {
            target[i] = source[i].toByte()
        }
        return target
    }

    fun size(): Int {
        return source.size
    }

}
